File: Beakjoon/_etc/Java/1699 제곱수의 합.java
import java.util.Scanner;

public class Main {
    public static void main(String args[]) {
      Scanner s = new Scanner(System.in);
      int n = s.nextInt();
      int[] dp = new int[100001];
      
      dp[0] = 0;
      dp[1] = 1;
      dp[2] = 2;
      dp[3] = 3;
      
      for(int i = 4; i <= n; i++) {
          
          if((int)Math.pow((int)(Math.pow(i,0.5)), 2) == i) dp[i] = 1;
          else {
              dp[i] = dp[i-1] + 1;
              for(int j = 1; j*j <= i; j++) {
                  dp[i] = Math.min(dp[i], dp[i-j*j] + dp[j*j]);
              }
              
              
          }
      }
      
      System.out.format("%d", dp[n]);
      
      
    }
}

// http://boj.kr/437f44e012bd4fccaf338469d22ba4a5

File: _etc/Kotlin/1699 제곱수의 합.kt
import java.util.Scanner
import kotlin.math.*

fun main(args: Array<String>) {
    val sc: Scanner = Scanner(System.`in`)
    val n = sc.nextInt()
    var dp = Array(100000){0}
    
    dp[0] = 0
    dp[1] = 1
    dp[2] = 2
    dp[3] = 3
    
    for(i in 4..n) {
        if((((i.toDouble().pow(0.5)).toInt()).toDouble().pow(2)).toInt() == i) dp[i] = 1
        
        else {
            dp[i] = dp[i-1] + 1
            var j = 1
            while(i > j*j) {
                dp[i] = min(dp[i], dp[i-j*j] + dp[j*j])
                j += 1
            }
        }
    }
    
    println(dp[n])
}

//http://boj.kr/6118174733b2410ebe63b370167ebd06
